lib/inferno/dsl/fhir_evaluation/rules/valuesets_demonstrate.rb

```
@@ -0,0 +1,207 @@
+              end
+            end
+
+            false
+          end
+
+          def extract_valueset(response)
+            value_set = JSON.parse(response.body)
+
+            if value_set['compose'] && value_set['compose']['include']
+              value_set['compose']['include'].map do |include|
+                include['concept']&.map { |concept| { system: include['system'], code: concept['code'] } }
+              end.flatten
+            else
+              puts 'No Value Set found in the response.'
+            end
+          end
+
+          def retrieve_valueset_api(url)
+            url['http:'] = 'https:' if url['http:']
+            uri = URI.parse(url)
+
+            http = Net::HTTP.new(uri.host, uri.port)
+            http.use_ssl = (uri.scheme == 'https')
+
+            request = Net::HTTP::Get.new(uri.request_uri)
+
+            username = config.data['Rule']['ValueSetsDemonstrate']['url']
+            password = config.data['Rule']['ValueSetsDemonstrate']['apikey']
+            encoded_credentials = Base64.strict_encode64("#{username}:#{password}")
+            request['Authorization'] = "Basic #{encoded_credentials}"
+
+            response = http.request(request)
+
+            content_type = response['content-type']
+            return unless content_type && !content_type.include?('text/html')
+
+            while response.is_a?(Net::HTTPRedirection)
+              redirect_url = response['location']
+
+              redirect_url['xml'] = 'json'
+              uri = URI.parse(redirect_url)
+
+              http = Net::HTTP.new(uri.host, uri.port)
+              http.use_ssl = (uri.scheme == 'https')
+
+              response = http.request(Net::HTTP::Get.new(uri.request_uri))
+            end
+
+            if response.code.to_i == 200
+              extract_valueset(response)
+            else
+              puts "Failed to retrieve the Value Set: #{url}. HTTP Status Code: #{response.code}"
```

I don't love this just being a line written to console, it's easy to miss. But I also don't know if this is worth being a full exception that halts all processing. I'm open to suggestions. Maybe raise an exception here and catch it up in `check`  - and add an error type message? Maybe it depends on how common we expect this to be. If truly exceptional maybe it's fine to let it halt all processing

---

Let me try to test with US Core versions to see how common they are and will come back with data.

---

So results from testing with US core vers are below:

It looks like failure to retrieve value sets in US Core isn't unusual.
My thought is theoretically if a value set isn't retrived, it shouldn't be included in the evaluation since we don't want to give any incorrect information to users. However we don't have to halt the whole evaluation because of that in this case: unlikely other rules, ValueSetsDemonstrate rule provides list of "unevaluated value sets" for those who are in the gray area. I am inclined to add failed retrieval into unevaluated with the reason of lack of information. Let me know what you think.
 

3.1.1 No error
4.0.0 No error

5.0.1
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.102.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.103.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.1.11.14914.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.836.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.837.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.877.
 
6.1.0
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.102.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.103.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1099.24.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1166.22.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.1.11.14914.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.836.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.837.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.877.
 
7.0.0
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.102.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.103.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1099.24.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1166.22.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.1.11.14914.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.836.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.837.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.877.
 
8.0.0
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.102.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1021.103.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1267.24.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.1.11.14914.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.836.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.837.
https://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.114222.4.11.877.



</body>

</html>

---

I just clicked on a few of these and they seem to load ok in a browser. Is it possible you hit a rate limiter or something?

But to answer your question, I guess I'm ok with considering these "unevaluated" but I would want to include some kind of additional detail on why it was unevaluated so the user can act appropriately

---

Ok I'm fine with that too. Another thought is, maybe have a config setting for this rule "ignore valuesets that can't be loaded" that if set, it just prints out an error message but otherwise ignores the exception. That way if there are reasons why it's common for VSes not to be reachable then the user isn't completely blocked, they can change the config

---

if the option "ignore valuesets that can't be loaded" = true
 -> Ignore the exception (no message) and proceed evaluation with the value set is considered as "unevaluated"

if the option "ignore valuesets that can't be loaded" = false
 -> Print out error message and halt the evaluation

Is this correct?

---

Revised 
https://github.com/inferno-framework/inferno-core/pull/636/commits/0bde8d57d9a92b4a76f76b3105bf82f68e0408b8